calour/ratio_experiment.py

```diff
@@ -0,0 +1,254 @@
+        store the data as sparse matrix (scipy.sparse.csr_matrix) or dense numpy array.
+    info : dict
+        information about the experiment (data md5, filenames, etc.)
+    description : str
+        name of the experiment
+    databases : dict
+        keys are the database names (i.e. 'dbbact' / 'gnps')
+        values are the database specific data for the experiment (i.e. annotations for dbbact)
+
+    See Also
+    --------
+    Experiment
+    '''
+    def __init__(self, *args, **kwargs):
+        super().__init__(*args, **kwargs)
+
+    def heatmap(self, *args, **kwargs):
+        '''Plot a heatmap for the ratio experiment.
+
+        This method accepts exactly the same parameters as input with
+        its parent class method and does exactly the sample plotting.
+
+        The only difference is that by default, it uses the diverging 'coolwarm' colormap and bad_color set to white.
+        You can always set it to other colormap/bad_color as explained in :meth:`.Experiment.heatmap`.
+
+        See Also
+        --------
+        Experiment.heatmap
+
+        '''
+        if 'cmap' not in kwargs:
+            kwargs['cmap'] = 'coolwarm'
+        if 'bad_color' not in kwargs:
+            kwargs['bad_color'] = 'white'
+        if 'clim' not in kwargs:
+            min_val = np.abs(np.nanmin(self.data))
+            max_val = np.abs(np.nanmax(self.data))
+            clim_val = np.max([max_val, min_val])
+            kwargs['clim'] = (-clim_val, clim_val)
+        super().heatmap(*args, **kwargs)
+
+    @classmethod
+    def from_exp(self, exp, common_field, group_field, value1, value2=None, threshold=5):
+        '''Create a RatioExperiment from two groups of samples in an experiment
+
+        Parameters
+        ----------
+        exp: calour.Experiment
+            The experiment to take the 2 groups of samples from.
+        common_field: str
+            Name of the sample metadata field to calculate the ratio within.
+            For example, to calculate the ratio between 2 timepoints for each individual,
+            common_field with be 'subject_id'
+        group_field: str
+            Name of the sample metadata field on which to divide to 2 groups for ratio calculation.
+            For example, to calculate the ratio between 2 timepoints for each individual,
+            group_field with be 'time_point'
+        value1 : str or iterable of str
+            Values for field that will be assigned to the first (nominator) sample group
+        value2: str or iterable of str or None, optional
+            If not None, values for field that will be assigned to the second (denominator) sample group.
+            If None, use all samples not in the first sample group as the second sample group.
+        threshold: float or None, optional
+            If not None, assign each data value<threshold to threshold. If both nominator and denominator are < threshold,
+            the resulting ratio will assigned be np.nan.
```

Comment:
is `nan` a good idea in data table? will it cause problems in other data manipulation functions? I haven't checked this - just brought it up.

Reply:
will check other functions to see they handle nan well.
nan is important since it means we do not know the ratio since the feature was not in the subset of samples...